framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+/****************************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <tinyara/audio/audio.h>
+#include <tinyara/audio/pcm.h>
+#include <tinyalsa/tinyalsa.h>
+#include <tinyara/fs/ioctl.h>
+#include <pthread.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sched.h>
+
+#include <mediaplayer/mediaplayer.h>
+
+/***
+ * Structure
+ */
+struct media_s {
+	dq_entry_t q;
+	int fd;
+	media_op_t op;
+	media_state_t state;
+	media_type_t type;
+	media_play_type_t play_type;
+};
+
+/***
+ * Static variables
+ */
+static pthread_t g_pth_playing;
+#ifdef CONFIG_AUDIO_MULTI_CARD
+static pthread_t g_pth_recording;
+#endif
+static sem_t g_sem_playing;
+static sem_t g_sem_recording;
+static pthread_mutex_t g_mutex_playing_q;
+static pthread_mutex_t g_mutex_recording_q;
+
+static bool g_playing_live;
+static bool g_recording_live;
+
+static dq_queue_t g_playing_q;
+static dq_queue_t g_recording_q;
+
+/***
+ * Pre-definitions
+ */
+media_result_t audio_playing(void *args);
+media_result_t audio_recording(void *args);
+
+media_result_t media_init(struct media_cb_s *cbs)
+{
+	pthread_attr_t attr;
+	pthread_mutexattr_t mutexAttr;
+	struct sched_param sparam;
+
+	// Init dq and sem
+	dq_init(&g_playing_q);
+	sem_init(&g_sem_playing, 0, 0);
+
+	pthread_mutexattr_init(&mutexAttr);
+	pthread_mutexattr_settype(&mutexAttr, PTHREAD_MUTEX_NORMAL);
+	pthread_mutex_init(&g_mutex_playing_q, &mutexAttr);
+
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
```

Comment:
Actually I think we'd better apply multi card as a default.(of course code related to this config should be removed)
Then we can use 2 thread as a default, code complexity will be decreased.
Of course singleton design concept should be applied.
Also each thread should work diffrently.
1. Play thread should store play position to resume again without h/w support.
    (Calculating file pos is needed based on sample rate, bit rate....and so on)
2. Record thread should call pcm_drain when pause or stop called.

Reply:
The current code supports both record and play in single and multi card enviroments.
Do you want me to remove support of single audio cards?
Currently both threads very similar. Just one does recording and one playback. Why do you need to have threads which work differently? .
Also the accuracy for pause and resume depends on tinyalsa layer. Currently, tinyalsa layer does not give any information about the amount of buffer that has been played. To get more accurate buffers we need to have small buffers so that pause and resume are more accurate. but some support is needed from tinyalsa to tell how many buffers were dropped from playback. Without it , it would be difficult to estimate which buffer has been played and which buffer has been dropped.

Reply:
Currently play & record is very simple so they are very simlar now.
But in future, their role will be more diffrent.
Also track is not role of tinyalsa, media player should calculate position of track.(mediaplayer already know seek point of file).
I want seprate it because I think none of h/w support pause & resume in play & record case.
And I think single buffer is enough. because we don't have to record and play at the same time.
In other words, save context of play, record at pause, and re-use saved context at resume.
we can seperate it to 2 thead for expandability, but we can share the buffer, that's what I'm saying.

Reply:
Currently context is saved but pause is not accurate. The play continues for some time after pause is called. This is due to the fact that currently we have not merged pcm_drain/pcm_drop.

Reply:
I estimate that pcm_drain/drop will be merged faster than this because mediaplayer is much bigger than it.
Please assume that code from @kishore-sn applied.